[1.18.x] Added a hook to activate spyglass-style scoping

Da-Technomancer opened the pull request:
This PR is a previous PR (https://github.com/MinecraftForge/MinecraftForge/pull/8171), rebased onto 1.18.

Added a hook to trigger the 'scoping' effect the spyglass uses (zoom in, visual overlay, adjusted movement).

Mods can register a predicate which will be tested on the client side to see whether the scoping effect is being externally triggered. If any registered predicate returns true, scoping will be active.

Also wrote a test mod.

Da-Technomancer committed:
Added a hook to activate spyglass-style scoping

Added a hook to trigger the 'scoping' effect the spyglass uses (zoom in, visual overlay, adjusted movement)
Mods can register a predicate which will be checked on the client side to see whether the scoping effect is being externally triggered
Also wrote a test mod.

Da-Technomancer committed:
Moved spyglass hooks to a dedicated class, applied feedback

*Moved the spyglass hooks from ForgeHooks to a dedicated class (SpyglassHooks). Reverted all changes made to ForgeHooks.
*Corrected several formatting issues
*Optimized the isPlayerScoping method
*Made the registerSpyglassCondition method synchronized, to make it threadsafe

Da-Technomancer committed:
Update ForgeHooks.java

Da-Technomancer committed:
Cleaned up formatting, javadocs, variable name

Da-Technomancer committed:
Reverted the auto-generated extraneous changes

Reverted the changed file that had an added header (generated when generating license headers), as not related to the purpose of this PR

Da-Technomancer committed:
Revert "Reverted the auto-generated extraneous changes"

This reverts commit 54e24a3c7a285a157d7a64eff974d939d1f471da.

Shadows-of-Fire commented:
Looks fine to me, would go ahead and assign but the CI is failing and I can't check if it compiles manually right now.

Da-Technomancer commented:
CI is saying it passes now. I didn't make any changes.

Shadows-of-Fire commented:
Yeah it just had to be restarted, figured that was the case.

diesieben07 commented on src/main/java/net/minecraftforge/common/SpyglassHooks.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.List;
+import java.util.function.Predicate;
+
+public class SpyglassHooks
```
This doesn't need to be its own class, you can use `ForgeHooks`.

diesieben07 commented on src/main/java/net/minecraftforge/common/SpyglassHooks.java:
```diff
@@ -0,0 +1,57 @@
+     *
+     * @param spyglassCondition A predicate which will be checked on the client side.
+     */
+    public static synchronized void registerSpyglassCondition(Predicate<Player> spyglassCondition)
```
This is not truly thread safe.
Please instead fire an event at startup to collect all conditions and then store them in an immutable and therefor threadsafe list.

Da-Technomancer replied:
Could you clarify why you consider it thread-unsafe? I must be missing something, but I don't see it.

diesieben07 replied:
`SPYGLASS_PREDICATES` is accessed from two methods. One of them is synchronized, one of them is not.
However that is irrelevant here, the more important change request is to add a dedicated collection event.

diesieben07 reviewed (changes requested):
Please update the license headers to 2022 as well and resolve merge conflicts.

Da-Technomancer committed:
Moved condition registration to registry, clean-up, merged with ForgeHooks

*As suggested, created a new registry for spyglass condition predicated, to make registration threadsafe. Removed the previous method for that purpose.
*As suggested, merged SpyglassHooks into ForgeHooks.
*Cleaned up comments.
*Fixed the spyglass hook not applying the zoom effect under all appropriate use cases (previously only applied while using any item).

Da-Technomancer committed:
Merge branch '1.18.x' into Spyglass-hook

diesieben07 reviewed (changes requested):
I don't think these should be a full blown registry. Why does it need to be?

Da-Technomancer commented:
> I don't think these should be a full blown registry. Why does it need to be?

It had seemed cleaner to use the tried-and-tested registry model than to create a new system for registration at mod startup. I'm open to suggestions.

diesieben07 commented:
Fire an event. That event has an internal list of predicates and a method `addPredicate`. After the event is fired, bake the list into an immutable one that is then stored and used later. This avoids threading issues and the need for synchronization, it improves performance, etc.

Da-Technomancer commented:
After several revisions, I'm calling it quits.
I was not able to find a way to fire an event only as needed which was both robust and cleanly implemented; I'm sure a way exists, but someone more familiar with the inner workings of Forge would need to find it. 

As a workaround, for anyone who found this PR because they wanted to add their version of a spyglass feature, here's how I suggest doing it without relying on this PR. It's only a rough approximation of what this PR is meant to achieve.
- Achieve the 'zoom' effect using an FOV modifier. There is an existing forge hook for FOV; see 'ForgeHooksClient::getFieldOfView' and set the FOV to 0.1 when zooming is desired. 
- Achieve the spyglass overlay effect using an event handler for the game rendering, and create your own overlay render.
- Achieve the sound by playing the sound (see the vanilla spyglass item class)
